[김은혜] Step3 구간 추가 기능 PR 요청

eunveloper opened the pull request:
안녕하세요 Step3 PR 요청드립니다!
요구사항을 구현 하면서 이것저것 API 호출 시, 상태 케이스를 생각하다보니 굉장히 까다롭고 어려웠습니다 ㅠㅠ
코드 구조 변경을 몇번이나 변경했는지...🤣🤯

기능을 구현하면서 개인적인 궁금증이 생겼습니다.
Service 레이어에서 Repository 레이어로 데이터베이스를 호출하는 지금 구조상
한개의 서비스에서 여러개의 레파지토리가 필요할때 구성방식이

1. 한개의 서비스에서 필요한 '레파지토리' 들을 주입하여 바로 데이터를 호출하여 사용한다.
2. 한개의 서비스에서 필요한 '서비스' 들을 주입하여 데이터를 호출하는 메소드를 사용한다.

둘중에 어떤방식을 써야할지 항상 고민인데요 (실무에서도요)
리뷰어님은 어떤 방식을 써야한다고 생각하시나요?

1번의 경우에는 레파지토리에서 데이터를 호출하는 부분이 중복될 가능성이 크고
2번의 경우에는 서비스끼리 의존성이 강하게 생기며 서로를 의존하는 양방향 의존관계가 생기게 될 가능성이 크다고 봅니다 ㅠㅠ

이번 미션에서는 1번의 방식으로 구현하였는데, 과연 어떤게 좀더 적절한 방법일까요?

eunveloper committed:
docs: README.md 작성

eunveloper committed:
docs: README.md 작성

- 인수테스트 시나리오 작성

eunveloper committed:
feat: 테스트코드 구성

- 구간 추가 기능 테스트코드 뼈대 구성
- 공통 테스트 픽쳐 유틸 클래스 분리
- CRUD 테스트 공통 함수 유틸 클래스 분리

eunveloper committed:
feat: Line <-> Station 연관관계 생성

- LineStation 구성
- LineStations 구성
- 연관관계 변경으로 인한 비지니스 로직 수정

eunveloper committed:
feat: 역 사이에 새로운 역을 등록할 경우

- 인수 시나리오 재정의
- 테스트 코드 작성
- 비지니스 로직 개발
- 인수 테스트 리팩터링
- 비지니스 로직 리팩터링

eunveloper committed:
feat: 새로운 역을 상행 / 하행 종점으로 등록할 경우

- 인수 시나리오 재정의
- 테스트 코드 작성
- 비지니스 로직 개발
- 인수 테스트 리팩터링
- 비지니스 로직 리팩터링

eunveloper committed:
feat: 각종 예외상황의 경우

- 인수 시나리오 재정의
- 테스트 코드 작성
- 비지니스 로직 개발
- 인수 테스트 리팩터링
- 비지니스 로직 리팩터링
- ExceptionHandler 추가

eunveloper committed:
refactor: 각종 리팩토링

- 메소드 변수명 변경
- 불필요 로직 제거
- 엔티티에게 메세지 주는 방식으로 함수 구현

geunhwanlee commented on README.md:
```diff
@@ -113,4 +113,77 @@
+- 인수 테스트의 결과가 다른 인수 테스트에 영향을 끼치지 않도록 인수 테스트를 서로 격리 시키세요.
+- 인수 테스트의 재사용성과 가독성, 그리고 빠른 테스트 의도 파악을 위해 인수 테스트를 리팩터링 하세요.
+
+## 인수테스트 조건명세
```
인수조건을 잘 정리해주셨네요! 👍🏻

geunhwanlee commented on README.md:
```diff
@@ -113,4 +113,77 @@
+- When 상행역과 하행역 사이의 새로운 역을 등록하면
+- Then 정상적으로 등록되고
+- Then 지하철 노선 조회시 3개의 지하철 역을 가지게 되고
+- Then 상행역 하행역 모두 변경이 없다.
```
이번 단계에서
`상행역`, `하행역`과
`상행종점역`, `하행종점역`은 다른 의미로 쓰입니다. 😀
```suggestion
- Then 상행종점역 하행종점역 모두 변경이 없다.
```
그러므로 이렇게 쓰는게 적절하지 않을까요?

eunveloper replied:
아.. 상행역 / 하행역은 해당 구간에 끝과 끝을 의미하고
상행종점역 / 히행종점역은 전체 호선의 끝과 끝을 의미하는 거군요 🤯

geunhwanlee commented on README.md:
```diff
@@ -113,4 +113,77 @@
+
+## 인수테스트 조건명세
+
+### 역 사이에 새로운 역을 등록할 경우
+- Given 상행역 하행역이 등록된 노선을 생성하고
+- When 상행역과 하행역 사이의 새로운 역을 등록하면
+- Then 정상적으로 등록되고
+- Then 지하철 노선 조회시 3개의 지하철 역을 가지게 되고
+- Then 상행역 하행역 모두 변경이 없다.
```
역과 역 사이에 새로운 역을 등록하게되면
기존 구간의 `상행역 `또는 `하행역`이 변경되고 역간 거리가 변경되는데요
정상적으로 변경됐는지 확인해보는것도 좋을것같아요! 😀

eunveloper replied:
아 거리를 변경하는 로직이 전체적으로 빠져있네요 ㅠㅠ 기능구현과 테스트코드 작성하겠습니다! 🙄

geunhwanlee commented on README.md:
```diff
@@ -113,4 +113,77 @@
+### 상행역과 하행역 둘 중 하나도 포함되어있지 않으면 추가할 수 없음
+- Given 상행역 하행역이 등록된 노선을 생성하고
+- When 상행역과 하행역과 동일한 구조의 새로운 역을 등록하면
+- Then 정상적으로 등록되지 않는다.
```
```suggestion
### 상행역과 하행역 둘 중 하나도 포함되어있지 않으면 추가할 수 없음
- Given 상행역 하행역이 등록된 노선을 생성하고
- When 상행역과 하행역이 모두 기존 등록된 역에 존재하지 않으면
- Then 정상적으로 등록되지 않는다.
```
인수조건이 잘못된것같네요! 😀

geunhwanlee commented on README.md:
```diff
@@ -113,4 +113,77 @@
+
+
+### 새로운 역을 상행 종점으로 등록할 경우
+- Given 상행역 하행역이 등록된 노선을 생성하고
```
`상행종점역 - 중간역 - 하행종점역`과 같이 두개 이상의 구간을 갖는 노선을 생성하고 테스트해보면 어떨까요? 😀

eunveloper replied:
기능구현을 할땐 이케이스까지 고려하느라 머리가 꽤나 아팠는데, 테스트로 구현을 못해봤네요 😂
테스트 구현해보고 실패하면 꽤나 당황스럽겠어요..ㅎㅎ 도전!

geunhwanlee commented on src/main/java/nextstep/subway/application/LineService.java:
```diff
@@ -20,20 +21,28 @@ public class LineService {
-        Line line = lineRepository.save(lineRequest.toLine(upStation, downStation));
-        return LineResponse.of(line);
+
+        LineStations lineStations = new LineStations();
+        lineStations.add(lineStationRepository.save(upStation.toLineUpStation()));
+        lineStations.add(lineStationRepository.save(downStation.toLineStation(upStation, lineRequest.getDistance())));
```
`null - upStation`
`upStation - downStation`
두개의 `lineStation`을 만들었는데요
`upStation - downStation` 만 만들어도 될것 같은데 두개를 만드신 이유가 있을까요? 🤔

eunveloper replied:
> 구간은 상행역과 하행역 거리를 모두 포함해야할것으로 보이는데
> 하행역만을 갖는 구간은 어떤 역할을 갖는 구간일까요? 🤔

이 피드백의 댓글의 답변과 동일합니다! 기준역과 앞선 상행역으로 엔티티가 구성되어있는 방식으로 구현이 되어있어요 ㅠㅠ!

geunhwanlee commented on src/main/java/nextstep/subway/application/LineService.java:
```diff
@@ -57,28 +66,18 @@ public LineResponse updateById(Long id, LineRequest lineRequest) {
     @Transactional
     public void deleteById(Long id) {
-        checkExistLine(lineRepository.findById(id), id);
+        getLine(id);
         lineRepository.deleteById(id);
```
```suggestion
        lineRepository.delete(getLine(id));
```
deleteById는 내부적으로 id를 가지고 엔티티를 조회하기 때문에
line을 조회했다면 이렇게 작성해도 될거같네요! 😀
```java
@Transactional
@Override
public void deleteById(ID id) {

Assert.notNull(id, ID_MUST_NOT_BE_NULL);

delete(findById(id).orElseThrow(() -> new EmptyResultDataAccessException(
		String.format("No %s entity with id %s exists!", entityInformation.getJavaType(), id), 1)));
}
```

eunveloper replied:
오호 이건 몰랐던 사실이네요! 내부적으로 select -> delete 구조이니 앞서 select를 했다면 그냥 delete만 할 수 있도록!

geunhwanlee commented on src/main/java/nextstep/subway/application/SectionService.java:
```diff
@@ -0,0 +1,64 @@
+        return LineResponse.of(line);
+    }
+
+    private void checkValidationParameter(SectionRequest sectionRequest) {
+        if (sectionRequest.hasNullValue()) {
+            throw new InvalidParameterException("상행역과 하행역이 모두 등록되어 있어야 합니다.");
+        }
+    }
```
`상행역`과 `하행역`을 검증하는 유효성검사는 `LineStation`에서 하는건 어떨까요? 😀

eunveloper replied:
제가 앞서 서비스 로직에서 구현했던 이유는, 기본적인 파라미터 널체크를 하기 위함이었는데요.
서비스 로직에서 레파지토리로 접근하는 비용을 쓰기전에
애초에 서비스 로직을 타지 않아도 되는 예외상황을 먼저 잡고자 했던거 같아요!

제가 저 기능을 착오한걸수도 있는데, 저는 API 요청시 상행역과 하행역을 모두 요청해야(파라미터에 담아야)한다라는
요구사항으로 이해해서 파라미터 널체크만 하거든요!

그래서 굳이 불필요한 비지니스 로직을 타기전에 예외처리를 해서 리턴하고자 했는데, 제가 잘못 접근한걸까요? 🥹

저 '등록' 이라는 용어가 데이터베이스에 등록이 되어있는 역으로만 가능하다 라는 요구사항이라면
`private Station getStation(Long id) {
        return stationRepository.findById(id)
                .orElseThrow(() -> new EntityNotFoundException("Station", id));
 }`
이라는 곳에서 예외처리가 될거같은데.. 이부분은 어떻게 생각하세요? 🤯🥹🤔

geunhwanlee replied:
API 요청 파라미터에 대한 필수값 검증으로 사용하신거라면 서비스로직을 타기전에 검증하는것도 좋습니다! 👍🏻
다만 현재 검증 로직이 서비스레이어에 있기때문에 도메인로직이라고 생각이 들어서 코멘트를 드렸던거에요. 😀

만약 API 파라미터의 필수값을 체크하는거라면 controller 또는 dto에서 검증할 수 있을거라고 생각하고
그것과 무관하게 도메인모델에서도 `상행역`과 `하행역`을 검증하면 좋겠네요! 😀

> 상행역과 하행역이 이미 노선에 모두 등록되어 있다면 추가할 수 없음
> 상행역과 하행역 둘 중 하나도 포함되어있지 않으면 추가할 수 없음

위 두가지 예외케이스도 함께 검증해볼수 있을것 같아요!

eunveloper replied:
서비스로직 -> DTO로 이동하도록 할게요!
추가로 도메인 모델에서 필요한 검증은 레파지토리에서 조회된 도메인 객체에서 진행하겠습니다 😆

geunhwanlee commented on src/main/java/nextstep/subway/domain/Line.java:
```diff
@@ -11,28 +11,18 @@ public class Line {
-    private int distance;
+
+    @Embedded
+    private LineStations lineStations = new LineStations();
```
일급컬렉션 👍🏻

geunhwanlee commented on src/main/java/nextstep/subway/domain/LineStation.java:
```diff
@@ -0,0 +1,61 @@
+
+import javax.persistence.*;
+
+@Entity
+public class LineStation {
```
미션에서 `상행역`과 `하행역` 그리고 `역간 거리`를 포함하는 개념으로 `구간`이라는 용어를 쓰고있고
API 스펙에서도 `section`이라는 용어를 쓰고있어요! 😀
`LineStation`보다는 `Section`이라는 명칭이 더 적합하지 않을까 생각되는데 어떻게 생각하시나요? 🤔

eunveloper replied:
그게 맞을거 같아요! 테스트 코드 클래스명도 Section이기도 하고...
제가 개발하다보니 곳곳에 LineStation과 Section이라는 명칭이 혼동되더라구요 ㅠㅠ 잡아주셔서 감사합니다!

geunhwanlee commented on src/main/java/nextstep/subway/domain/LineStation.java:
```diff
@@ -0,0 +1,61 @@
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+
+    @ManyToOne(fetch = FetchType.LAZY)
+    @JoinColumn(name = "preStationId")
+    private Station preStation;
+
+    @ManyToOne(fetch = FetchType.LAZY)
+    @JoinColumn(name = "stationId")
+    private Station station;
```
API 스펙에서 `상행역`은 `upStation` `하행역`은 `downStation`이라는 명칭을 쓰고있는데
`preStation`, `station`이라는 명칭보다 `upStation`, `downStation`이라는 명칭이 더 적합하지 않을까요? 🤔

eunveloper replied:
> 구간은 상행역과 하행역 거리를 모두 포함해야할것으로 보이는데
> 하행역만을 갖는 구간은 어떤 역할을 갖는 구간일까요? 🤔

이 피드백의 댓글의 답변과 동일합니다! 기준역과 앞선 상행역으로 엔티티가 구성되어있는 방식으로 구현이 되어있어요 ㅠㅠ!

geunhwanlee commented on src/main/java/nextstep/subway/domain/LineStation.java:
```diff
@@ -0,0 +1,61 @@
+
+    protected LineStation() {}
+
+    public LineStation(Station station) {
+        this.station = station;
+    }
```
`구간`은 `상행역`과 `하행역` `거리`를 모두 포함해야할것으로 보이는데
`하행역만을 갖는 구간`은 어떤 역할을 갖는 `구간`일까요? 🤔

eunveloper replied:
음 제가 설계를 좀 다른방향으로 하게되서 생긴 이슈인거 같은데요,
A - B - C 라는 구간이 있으면 (A-B) (B-C) 이 두가지 구간만 있으면 되는걸까요?

저는 (null - A) (A-B) (B-C) 라는 구간을 생각했어요! 위에 댓글과 동일한 맥락인데
구간이라는 엔티티를 정할때 메인 역과 그 메인의 앞선역 두가지로 구간을 관리해야한다고 생각해서
필드명도 station (메인역) preStation(앞선역) 으로 지정했던 거였거든요!

이걸 통해서 해당 호선에 제일 앞선  구간을 파악하고 내부 로직을 통해서 호선내 구간을 정렬하여 리턴하게 구성해서요!

예를들면...
A-B-C-D 가있는 라인을 호출하면서 반환된 리스트들이
(A-B)
(C-D)
(null-A)
(B-C) 다음과 같다면

(null-A)를 기준으로 호선 순서를 정렬하여 리턴해 주게 되어있어요!
제가 요구사항을 잘못 접근한걸까요...? 🤣

geunhwanlee replied:
이 질문은 DM으로도 주셔서 DM의 답변을 인용할게요!

> 정렬의 시작점을 표현하기 위해 null을 사용하신것은 좋습니다!
>
> 다만 정렬을 위해 도메인 모델인 구간의 구조를 변경했다고 개인적으로 생각이 들어요.
> 제가 코드를 읽으면서 복잡하게 느꼈던 부분도 이러한 이유때문일거같아요.
> 저는 이를 로직이 섞여있다 또는 로직이 숨어있다 라고 생각하고
> 로직을 드러내주는것이 좋다고 생각해요!
>
> 이 케이스에서 예를들면 정렬을 위해 null-* 구간을 유지해야하고
> 그것을 유지하기 위해 정렬과 무관한 메서드에서도 null-* 구간을 유지하기 위한 로직들이 들어가게 될거에요.
> 내가 아닌 다른사람이 코드를 변경했을때 null-* 구조가 깨질 가능성도 있구요!
>
> 그러므로 저는 구간은 정렬에 관심을 두지 않고 구간을 표현하는데만 관심을 둬야한다고 생각하고
> 정렬하는 로직은 따로 드러내서 구간 리스트를 받아 지하철역의 리스트로 정렬해주는 역할만을 하는게 좋지않을까 라고 생각이 드네요!

geunhwanlee commented on src/main/java/nextstep/subway/domain/LineStation.java:
```diff
@@ -0,0 +1,61 @@
+        this.distance = distance;
+    }
+
+    public void resetStation(Station station) {
+        this.station = station;
+    }
+
+    public void resetPreStation(Station station) {
+        this.preStation = station;
+    }
```
`구간`의 `상행역` 또는 `하행역`이 변경되면 `역간 거리`도 변경되어야할것 같은데 어떻게 생각하시나요? 😀

eunveloper replied:
이건 제가 기능을 누락했네요! 추가하겠습니다 ㅎㅎ

geunhwanlee commented on src/main/java/nextstep/subway/domain/LineStation.java:
```diff
@@ -0,0 +1,61 @@
+    }
+
+    public boolean isGatherThanDistance(LineStation infixLineStation) {
+        if (this.distance == 0) {
+            return false;
+        }
```
`역간 거리`가 0이 되는 케이스가 있을까요? 🤔
`역간 거리`가 0이될때 예외처리를 한다면 `역간 거리`를 0보다 큰 값으로 보장할수 있지 않을까요?

eunveloper replied:
> 구간은 상행역과 하행역 거리를 모두 포함해야할것으로 보이는데
> 하행역만을 갖는 구간은 어떤 역할을 갖는 구간일까요? 🤔

이 피드백의 댓글의 답변과 동일합니다! 기준역과 앞선 상행역으로 엔티티가 구성되어있는 방식으로 구현이 되어있어서 상행역이 없는 구간이라면 역간 거리가 0으로 계산된다고 생각하였습니다...!!

geunhwanlee commented on src/main/java/nextstep/subway/domain/LineStation.java:
```diff
@@ -0,0 +1,61 @@
+        if (this.distance == 0) {
+            return false;
+        }
+        return infixLineStation.distance >= this.distance;
```
`LineStation.isGraterThanDistance(infixLineStation);`의 메서드명을 보면
`LineStation`이 `infixLineStation`보다 거리가 긴지 확인하는 메서드로 읽혀져요!
하지만 부등호의 방향은 반대네요. 😭
적절한 메서드명으로 변경해야할것 같아요.

eunveloper replied:
오호 그렇게 읽히네요...@.@ lessThan이 더 적합할거 같아요 ㅎㅎ

geunhwanlee commented on src/main/java/nextstep/subway/domain/LineStation.java:
```diff
@@ -0,0 +1,61 @@
+        return infixLineStation.distance >= this.distance;
+    }
+
+    public boolean isEqualsId(Long id) {
+        return station.getId().equals(id);
+    }
```
`하행선`의 id와 같은지 체크하는 메서드네요!
더 적절한 메서드명으로 변경하는게 좋을것같아요. 👍🏻

eunveloper replied:
하행선의 id로만 제한하지 않고 어떤 아이디든 비교하는 메서드로 사용하려고 했는데...
구현하다보니 어째 하행선 id랑만 비교하는 메서드가 되었네요 😂

geunhwanlee commented on src/main/java/nextstep/subway/domain/LineStation.java:
```diff
@@ -0,0 +1,61 @@
+        return station.getId().equals(id);
+    }
+
+    public Station getStation() {
+        return station;
+    }
+
+    public Station getPreStation() {
+        return preStation;
+    }
```
도메인모델에서 getter는 지양하는게 좋습니다!
getter를 사용하지 않고 도메인모델에게 적절한 메시지를 보내서 처리하면 어떨까요? 🙏

eunveloper replied:
최대한 제거하는 방향으로 진행했습니다! 하지만 몇개 남는건 어쩔수 없나봐요 😂

geunhwanlee commented on src/main/java/nextstep/subway/domain/LineStations.java:
```diff
@@ -0,0 +1,111 @@
+import java.util.ArrayList;
+import java.util.List;
+
+@Embeddable
+public class LineStations {
```
👍🏻

eunveloper replied:
이것도 동일하게 Sections 로 이름 변경하였습니다 😆

geunhwanlee commented on src/main/java/nextstep/subway/domain/LineStations.java:
```diff
@@ -0,0 +1,111 @@
+@Embeddable
+public class LineStations {
+
+    @OneToMany(fetch = FetchType.LAZY)
+    @JoinColumn(name = "line_id")
+    private List<LineStation> lineStations = new ArrayList<>();
```
`Line`과 `LineStation`은 OneToMany 단방향 관계로 보이네요!
OneToMany 단방향 관계는 어떤 문제점이 있을까요? 🤔

eunveloper replied:
OneToMany 단방향 관계는 연관관계 주인이 정확히 명시되어 있지 않아서
Insert나 delete할때 추가 update 쿼리가 나가는 방식으로 데이터 셋을 하는걸로 알고있어요!

양방향 관계로 변경하고 데이터 셋 구조를 변경해야 겠네요!! 🤔

geunhwanlee commented on src/main/java/nextstep/subway/domain/LineStations.java:
```diff
@@ -0,0 +1,111 @@
+    public void infixSection(LineStation infixLineStation) {
+        LineStation needChangedStation = getNeedChangedStationStation(infixLineStation);
+
+        if (needChangedStation == null) {
+            add(infixLineStation);
+            return;
+        }
```
`needChangedStation`이 null이라는게 어떤걸 의미하는지 드러나지 않네요.
`infixLineStation`이 `상행종점역` 또는 `하행종점역`을 등록하는 경우인것 같은데 적절한 명칭으로 변경하는게 좋지 않을까 싶어요! 🙏

eunveloper replied:
기능 전면 수정하면서 제거된 로직입니다 ㅎㅎ!

geunhwanlee commented on src/main/java/nextstep/subway/domain/LineStations.java:
```diff
@@ -0,0 +1,111 @@
+        }
+        checkValidationParameter(needChangedStation, infixLineStation);
+
+        if (needChangedStation.isEqualsId(infixLineStation.getStation().getId())) {
+            infixLineStation.resetStation(infixLineStation.getPreStation());
+            infixLineStation.resetPreStation(needChangedStation.getPreStation());
+        }
```
`기존 구간 사이에 새로운 역을 등록하는 경우`이고 `기존 구간의 하행선과 등록하려는 구간의 하행선이 일치하는 경우`에
`기존구간의 하행선`은 `등록하려는 구간의 상행선`으로 변경되고 `기존구간의 거리`는 `등록하려는 구간의 거리`만큼 감소되어야겠죠.
```suggestion
        if (needChangedStation.isEqualsId(infixLineStation.getStation().getId())) {
            needChangedStation.resetStation(infixLineStation.getPreStation());
            needChangedStation.거리감소(infixLineStation.getDistance());
        }
```
이렇게 되어야하지 않을까 싶네요! 😀

그리고 이렇게 하는것보다
`needChangedStation.fix(infixLineStation);`처럼 메시지를 던지면 적절히 역할을 위임할수 있을것같은데 어떻게 생각하시나요?

eunveloper replied:
네 맞아요! 저 if문 안에서 변경이 필요한 station에서 station 정보 + 거리정보를 모두 변경해야 하니,
각각의 메소드를 호출하는거보다 하나의 메소드로 메세지를 던져서 변경이 필요한 모든 정보를 한번에 변경하는 것이 좋을 것 같아요!

geunhwanlee commented on src/main/java/nextstep/subway/domain/LineStations.java:
```diff
@@ -0,0 +1,111 @@
+
+    public void infixSection(LineStation infixLineStation) {
+        LineStation needChangedStation = getNeedChangedStationStation(infixLineStation);
+
```
구간을 추가하기 전에

> 상행역과 하행역이 이미 노선에 모두 등록되어 있다면 추가할 수 없음
> 상행역과 하행역 둘 중 하나도 포함되어있지 않으면 추가할 수 없음

두가지의 예외케이스를 처리해야하지 않을까요? 🤔

geunhwanlee commented on src/main/java/nextstep/subway/domain/LineStations.java:
```diff
@@ -0,0 +1,111 @@
+    }
+
+    private void checkValidationParameter(LineStation existedStation, LineStation infixLineStation) {
+        if (existedStation.isEqualsId(infixLineStation.getPreStation().getId())) {
+            throw new InvalidParameterException("상행선과 하행선을 모두 동일하게 등록할 수 없습니다.");
+        }
```
`기존 구간의 하행선`과 `등록하려는 구간의 상행선`이 동일한것이 예외상황이 맞을까요? 🤔

eunveloper replied:
`상행역과 하행역이 모두 노선에 등록` 이 경우인데 메세지가 착오가 있을거 같네요 🤣

geunhwanlee commented on src/main/java/nextstep/subway/domain/LineStations.java:
```diff
@@ -0,0 +1,111 @@
+            throw new InvalidParameterException("상행선과 하행선을 모두 동일하게 등록할 수 없습니다.");
+        }
+
+        if (existedStation.isGatherThanDistance(infixLineStation)) {
+            throw new InvalidParameterException("기존의 역 사이보다 더 긴 길이의 역을 등록할 수 없습니다.");
+        }
```
해당 예외처리는 적절한 도메인객체에게 위임하는것이 어떨까요? 😀

eunveloper replied:
구간의 길이를 비교하는 거니! 구간 도메인에게 위임해야겠네요! 💪🏻

geunhwanlee commented on src/main/java/nextstep/subway/exception/GeneralExceptionHandler.java:
```diff
@@ -0,0 +1,23 @@
+        return newResponse(e, HttpStatus.BAD_REQUEST);
+    }
+
+}
```
공통 예외처리 👍🏻

geunhwanlee commented on src/test/java/nextstep/subway/line/LineAcceptanceTest.java:
```diff
@@ -1,41 +1,34 @@
 
     @Autowired
-    private StationRepository stationRepository;
+    private InitializationEntity initializationEntity;
+
+    @Autowired
+    private ExecuteRestEntity executeRestEntity;
```
테스트 픽스처를 생성하는 `InitializationEntity`와
RestApi를 실행하는 `ExecuteRestEntity`를 재사용 가능하도록 분리해주셨네요. 👍🏻

다만 `InitializationEntity`와 `ExecuteRestEntity`는 `Line`과 `Section` 두 테스트를 모두 포함하고있고
패키지 위치로 보건데 다른 테스트도 포함할 가능성이 있어보이네요.
적절히 분리하는게 좋을것같은데 어떻게 생각하시나요? 😀

eunveloper replied:
ExecuteRestEntity 이 클래스는 insert() select() 를 한번더 메소드화 해서 
전역적으로 Rest 요청이 있을때 location 값을 사용하도록 공통화 해놔서 괜찮다고 판단했고

InitializationEntity는 전체 엔티티들을 save() 해주고 있지만
테스트 클래스에서 적절히 필요한 엔티티들만 save() 하는 함수만 호출해서 데이터를 세팅하기 때문에 괜찮다고 판단했는데

리뷰어님은 어떤 방향이 좀더 나을꺼라고 생각하시나요?!

geunhwanlee replied:
공통적으로 사용하는 `insert`, `select` 등의 메서드들이 있네요!
그렇다면 공통적으로 사용할수 있는 메서드만 남겨두고
`insertLineSuccess`, `insertSectionSuccess`와 같은 메서드들은
`LineExecuteRestEntity`, `SectionExecuteRestEntity`와 같이 도메인별로 분리하는건 어떨까요? 😀
더 나은 방법이 있다면 그것도 좋습니다 👍🏻

`InitializationEntity`의 경우는 `initStation`, `initLine`, `initSection`이 독립적인 메서드로 분리되어있지만 각 메서드는 서로 의존성이 있네요.
`initSection`을 하기위해선 `initStation`, `initLine`이 먼저 선행되어야하기 때문에 독립적으로 실행될수 없네요. 😭
그리고 테스트에서 만약 `InitializationEntity.section_1` 데이터를 사용해서 테스트를 한다고 하면 `section_1`이 어떤 `line`의 구간이고 어떤 역을 포함하고있는지를 알기가 힘든 문제가 있습니다. 테스트데이터의 구조를 파악하기위해 `InitializationEntity`를 들여다 봐야하네요. 🤔

개인적으로 `Station`같은 독립적인 데이터들은 전역적으로 사용해도 괜찮다고 생각이 들어요.
다만 `Line`이나 `Section`과 같은 데이터들은 테스트 과정에서 그 내부 구조에 관심을 가질수밖에 없기 때문에 테스트코드 안에서 생성하고 그 구조를 드러내줘야한다고 생각해요.

예를 들면
```
// given
line = 노선등록되어있음("1호선", station1, station2);

// when
구간추가(line, station2, station3);
line = 노선조회(line.location);

// then
노선에포함되어있음(line, station1, station2, station3);
```
이런 형태로 테스트를 읽었을때 흐름이 보이도록 작성하는게 좋다고 생각해요.

geunhwanlee commented on src/main/java/nextstep/subway/domain/LineStations.java:
```diff
@@ -0,0 +1,111 @@
+@Embeddable
+public class LineStations {
+
+    @OneToMany(fetch = FetchType.LAZY)
+    @JoinColumn(name = "line_id")
+    private List<LineStation> lineStations = new ArrayList<>();
```
`구간`은 `노선`에 종속적인 관계라고 생각하는데요
`노선`이 삭제된다면 `연관된 구간들`은 어떻게 되어야할까요? 😀

eunveloper replied:
노선이 제거되면 구간은 지워져야 하지 않을까요? 🤔
구간에 합이 노선이니까..?

대신 역은 지워져선 안되겠어요!

geunhwanlee reviewed (changes requested):
3단계 미션 어려웠을텐데 잘 구현해주셨어요! 👍🏻

이번 단계는 난이도가 갑자기 올라가는 경향이 있어요. 😂
코멘트를 드렸는데 아무래도 이번 단계가 조금 어렵다보니
반영하는데 어려움이 있거나 궁금한게 있으시다면 언제든 DM 주시거나 질문 남겨주셔도 좋습니다!

그럼 화이팅 하세요! 🔥

\+
> 기능을 구현하면서 개인적인 궁금증이 생겼습니다.
> Service 레이어에서 Repository 레이어로 데이터베이스를 호출하는 지금 구조상
> 한개의 서비스에서 여러개의 레파지토리가 필요할때 구성방식이
> 
> 한개의 서비스에서 필요한 '레파지토리' 들을 주입하여 바로 데이터를 호출하여 사용한다.
> 한개의 서비스에서 필요한 '서비스' 들을 주입하여 데이터를 호출하는 메소드를 사용한다.
> 둘중에 어떤방식을 써야할지 항상 고민인데요 (실무에서도요)
> 리뷰어님은 어떤 방식을 써야한다고 생각하시나요?
> 
> 1번의 경우에는 레파지토리에서 데이터를 호출하는 부분이 중복될 가능성이 크고
> 2번의 경우에는 서비스끼리 의존성이 강하게 생기며 서로를 의존하는 양방향 의존관계가 생기게 될 가능성이 크다고 봅니다 ㅠㅠ
> 
> 이번 미션에서는 1번의 방식으로 구현하였는데, 과연 어떤게 좀더 적절한 방법일까요?

좋은 질문 해주셨어요! 👍🏻
은혜님 말씀이 맞습니다. 서비스에서 서비스를 의존하게되면 순환참조가 형성될 가능성이 있습니다!

저같은 경우는 `Repository`를 주입하여 사용하는편이에요!
데이터를 조회하는것 자체는 로직이 없기때문에 중복이라 볼수 없다고 생각하고
조회 후 결과가 empty일때 예외를 발생시키는부분이 중복된다고 볼수도 있지만
예외를 던지는것도 그 도메인의 비즈니스 로직이라고 본다면 중복이 아니라고 생각해요! 😀

eunveloper committed:
refactor: 피드백 반영

- 인수조건 작성중 혼동단어 변경
- deleteById -> delete 변경 (내부에서 조회로직이 포함됨)
- 클래스명 변경 (LineStation -> Section)

eunveloper committed:
refactor: 피드백 반영

- 인수조건 작성중 혼동단어 변경
- deleteById -> delete 변경 (내부에서 조회로직이 포함됨)
- 클래스명 변경 (LineStation -> Section)

eunveloper committed:
refactor: 각총 명칭 및 변수명 변경

- lineStation -> Section
- 상행역, 상행종점역 / 하행형, 하행종점역 구분

eunveloper committed:
refactor: 요구사항 기능 전면변경

- null-* 구간 제거
- 요청한 구간 조건에 맞춰 기존 구간 변경
- 기존 구간의 거리 변경

eunveloper committed:
refactor: 피드백 반영

- 단방향 OneToMany 양방향으로 변경
- 매핑방향 변경하면서 데이터 셋 로직 변경
- 비지니스 로직 리팩토링
- 테스트 코드 리팩토링
- 도메인 모델 getter 최대한 제거

eunveloper commented:
다시 PR 올렸습니다!
요구사항을 잘못된 방향으로 접근해서 굉장히 많은 부분이 바뀌어서... 다시 봐주셔야 하는 고생이 있으시겠어요 ㅠㅠ 죄송합니다

그래도 저번 PR에서 말씀해주신 피드백들 최대한 반영해서 재 구현 해보았습니다!
감사합니다 :)

geunhwanlee commented on src/main/java/nextstep/subway/application/LineService.java:
```diff
@@ -20,65 +21,62 @@ public class LineService {
-        return LineResponse.of(line);
+
+        Sections sections = new Sections();
+        sections.add(sectionRepository.save(
+                upStation.toSection(downStation, lineRequest.getDistance())));
```
`Section`의 `upStation`과 `downStation`이 반대로 저장되었네요.😭

`upStation.toSection(downStation, lineRequest.getDistance())` 코드를 보면 `upStation`과 `downStation`이 혼동될수 있을것 같은데요
`new Section(upStation, downStation, lineRequest.getDistance());` 이렇게 써도 괜찮을거같아요! 😀

eunveloper replied:
앗 그러네요 과거 구조가 아직 남아있었군요 ㅠㅠ 수정하겠습니다 ㅎㅎ!

geunhwanlee commented on src/main/java/nextstep/subway/application/LineService.java:
```diff
@@ -20,65 +21,62 @@ public class LineService {
-        return LineResponse.of(line);
+
+        Sections sections = new Sections();
+        sections.add(sectionRepository.save(
```
`Line`과 `Section`의 연관관계에 cascade PERSIST 옵션을 주면 `sectionRepository.save`처럼 각 엔티티들을 각각 저장할 필요가 없을것 같아요! 😀

cascade 옵션에는 그 외 DELETE, MERGE 등 여러 옵션이 있는데요
각각이 어떻게 동작하는지
그리고 `Line-Section` 관계에선 어떤 옵션이 적합한지 고민해보시면 좋을것같아요! 👍🏻

eunveloper replied:
cascade 옵션을 처음 접해봐서 제대로 활용을 못했네요 ㅠㅠ!
Line을 저장할때 Line 도메인에 Sections 객체를 잘 담아만 준다면!
영속성 전이를 통해서 Line과 함께 Section들도 저장되는걸 테스트 코드로 방금 확인했어요!

좀더 코드도 간결해지고 연관관계의 데이터 셋의 고민을 덜수 있겠네요!
Line을 생성할때 Section도 함께 등록해줘야하고
Line이 삭제된다면 그에 종속되는 Section들도 제거가 되야하니 cascade type은 ALL 이 좋을까요?!

geunhwanlee replied:
네 좋습니다 👍🏻

geunhwanlee commented on src/main/java/nextstep/subway/domain/Section.java:
```diff
@@ -0,0 +1,119 @@
+
+    protected Section() {}
+
+    public Section(Station upStation, Station downStation, int distance) {
+        this.upStation = upStation;
+        this.downStation = downStation;
+        this.distance = distance;
+    }
```
Section은 Line에 종속적인데
생성자에 Line을 포함해서 Line이 필수임을 보장해주는건 어떤가요? 😀

eunveloper replied:
위처럼 영속성 전이를 사용하지 못해서 생겨진 생성자네요!
위 방식으로 cascade를 사용하니 저 생성자를 사용하는 코드가 모두 제거되었어요... (편-안)

eunveloper replied:
제가 매핑관계에서 연관관계 주인에 대한 개념이 좀 부족했던거 같아요 😂 지금은 속이 좀 시원해졌어요
재가 매주 JPA 스터디를 하는데 다음차수부터 매핑관계에 대해서 나가는데 열심히 공부해봐야겠어요!

geunhwanlee commented on src/main/java/nextstep/subway/application/LineService.java:
```diff
@@ -20,65 +21,62 @@ public class LineService {
+
+        Sections sections = new Sections();
+        sections.add(sectionRepository.save(
+                upStation.toSection(downStation, lineRequest.getDistance())));
```
`Section` 생성시 `Line`이 포함되어있지 않아서 `Section`의 `line_id`가 null로 들어가고있네요. 😭

엔티티의 연관관계에서는 `연관관계의 주인`에 반드시 관계를 주입시켜줘야해요.
`Line-Section` 관계에서는 `Section`이 `line_id`를 가지고있기 때문에 `Section`이 연관관계의 주인이에요.
따라서 `Section-Line` 관계를 형성하려면 `Section`의 `Line`에 반드시 값이 있어야해요.

eunveloper replied:
이것도 마찬가지로 위의 댓글들처럼
연관관계의 주인에 대한 매핑의 개념이 부족했고 그로인한 영속성 전이가 부족해서 생긴 코드네요 ㅠㅠ
위의 댓글들 반영하니 말끔하게 사라지네요 ㅎㅎ 신기해요!

geunhwanlee commented on src/main/java/nextstep/subway/application/LineService.java:
```diff
@@ -20,65 +21,62 @@ public class LineService {
         return lineRepository.findAll()
                 .stream()
-                .map(LineResponse::of)
+                .map(Line::toLineResponse)
```
`LineResponse.of`에서 `Line.toLineResponse`로 변경하신 이유가 있을까요? 🤔

엔티티에서 DTO를 의존하는것보다 DTO에서 엔티티를 의존하는게 더 낫다고 생각해요. 😀
하나의 엔티티에서 여러 DTO가 나올수도 있기 때문에 DTO가 추가될때마다 엔티티를 변경하는건 바람직하지 않다고 생각해요.
그리고 DTO는 view에 연관이 있는데 엔티티에서 DTO를 의존하면 의존의 방향이 반대가 되는경향도 있습니다.

eunveloper replied:
아 도메인 모델의 getter 를 지양하는 부분을 반영하다보니 대부분의 getter가 DTO의 of 함수에서 사용중이더라구요!
그래서 의존 방향을 변경했는데
`하나의 엔티티에서 여러 DTO가 나올수도 있기 때문에 DTO가 추가될때마다 엔티티를 변경하는건 바람직하지 않다고 생각`
이 이야기를 들으니 방향이 잘못되긴 한거 같네요 ㅠㅠ

그럼 이경우에 사용되는 getter는 괜찮은 걸까요?

geunhwanlee replied:
도메인모델에서 getter를 지양하라는건
비즈니스 로직을 구현할때 getter를 통해 값을 꺼내온 후 도메인객체 외부에서 로직을 구현하지 말고
도메인객체에게 역할을 위임하여 구현하도록 하기위한 것으로 저는 이해하고있어요! 😀

그렇기 때문에 DTO로 값을 매핑하는 경우는 getter를 사용해도 무방하지 않을까 생각해요. 🤔

정답은 없으니 한번 생각해보시고 은혜님만의 답을 찾아가시면 좋을것 같아요! 👍🏻

geunhwanlee commented on src/main/java/nextstep/subway/application/LineService.java:
```diff
@@ -20,65 +21,62 @@ public class LineService {
     public LineResponse updateById(Long id, LineRequest lineRequest) {
         Line line = getLine(id);
-        return LineResponse.of(line.updateInfo(lineRequest));
+        return line.updateByLineRequest(lineRequest).toLineResponse();
```
```suggestion
        return line.update(lineRequest.toLine(null)).toLineResponse();
```
이렇게 하면 `Line`에서 `LineRequest`를 의존하는걸 피할수 있을것 같은데 어떻게 생각하시나요? 😀

eunveloper replied:
위와 같은 상황이네요!? 도메인 모델이 DTO를 의존하는 방향을 지양하자는 말씀이신거죠?! 🤔

geunhwanlee commented on src/test/java/nextstep/subway/section/SectionAcceptanceTest.java:
```diff
@@ -0,0 +1,187 @@
+        assertThat(response.statusCode()).isEqualTo(HttpStatus.BAD_REQUEST.value());
+    }
+
+    /**
+     * When 상행역과 하행역과 동일한 구조의 새로운 역을 등록하면
+     * Then 정상적으로 등록되지 않는다.
+     */
+    @DisplayName("상행역과 하행역이 이미 노선에 모두 등록되어 있다면 등록할 수 없다.")
+    @Test
+    void createFailSectionSameStation() {
+        // when
+        ExtractableResponse<Response> response = executeRestEntity
+                .insert(executeRestEntity.generateSectionRequest(
+                        upStation.getId(), mediumStation.getId(), LESS_THAN_DISTANCE), INSERT_SECTION_LOCATION).extract();
+
+        // then
+        assertThat(response.statusCode()).isEqualTo(HttpStatus.BAD_REQUEST.value());
+    }
```
`upStation-mediumStation-downStation` 의 구조를 갖는 `노선`에서
`upStation-downStation` 또는 `downStation-upStation` 또는 `mediumStation-upStation` 등의 `구간`을 등록한다면 등록할수 없어야해요.

하지만 등록이 되네요.. 😭
한번 확인해보시고 수정해주세요. 🙏

eunveloper replied:
테스트케이스가 좀 부족했나봐요 ㅠㅠ 좀더 보강하겠습니다! 😂

geunhwanlee commented on src/main/java/nextstep/subway/domain/Sections.java:
```diff
@@ -0,0 +1,95 @@
+    private Section getInnerStationSection(Section infixSection) {
+        Section changedStationSection = sections
+                .stream()
+                .filter(section -> section.isUpStationEqualsId(infixSection.getUpStation().getId()))
```
```suggestion
                .filter(section -> section.isSameUpStation(infixSection))
```
section에 private로 정의해둔 메서드를 활용하면 getter를 사용하지 않고 처리할수 있을것 같네요! 👍🏻

eunveloper replied:
앗 제가 만들어놓고...😅

eunveloper replied:
수정하고 테스트코드 돌려보니 하는 행위가 좀 다르네요 ㅠㅠ
isUpStationEqualsId() 는 section의 upstation의 id가 파라미터로 넘어온 (upstation인지 downstation인지 알수 없음) id값과 동일한건지 알아보는거고
isSameUpStation() 는 section의 upstation과 파라미터로 넘어온 section의 upstation이 동일한지 확인하네요!

eunveloper replied:
다만 getter를 좀 지양하는 방법이 있는지 한번 고민해보겠습니당!

geunhwanlee commented on src/main/java/nextstep/subway/domain/Station.java:
```diff
@@ -18,12 +20,20 @@ public Station(String name) {
 
-    public String getName() {
-        return name;
+    public boolean equalsById(Station station) {
+        return id.equals(station.getId());
+    }
```
엔티티의 id가 같다는건 두 엔티티가 동등하다는 의미로 봐도 되지 않을까요? 🤔

`equals`와 `hashCode`를 재정의하여 id가 같으면 같은 엔티티라고 표현해주는것도 좋을것 같아요!

그럼 두 `Station`을 비교할때 `equals`를 통해 비교할수 있겠네요.

eunveloper replied:
음 저도 이부분은 고민을 해보기도하고 이번에 스터디 하면서도 좀 알아봤는데,
equals와 hashCode에 단순 id만 비교하게 오버라이딩을 하면 id가 없는상태 (비영속상태) 에서 문제가 생길 수 있는 여지가 다분하더라구요...!

그래서 그냥 id로만 단순비교하는 메소드를 따로 구성했어요! 말그대로 아이디가 같은지 비교하자는 의도로요!
근데 다만 저 id.equlas()에서 NPE 방지코드가 필요할거 같긴 하네요 ㅠㅠ........

geunhwanlee commented on src/main/java/nextstep/subway/dto/LineResponse.java:
```diff
@@ -1,31 +1,25 @@
-        return new LineResponse(line.getId(), line.getName(), line.getColor(),
-                line.getUpStation(), line.getDownStation());
-    }
+    private Set<StationResponse> stations = new HashSet<>();
```
노선에 포함된 역이 중복되지 않는다는 특성을 `Set` 자료구조로 표현해주신점 좋습니다. 👍🏻

한가지 확인하실 부분은 `Set`의 구현체가 `HashSet`인데요
`HashSet`은 역의 순서를 보장하지 않습니다! 😀

그렇다면 역의 순서를 정렬하기 위해선 어떤 자료구조를 써야할까요?

eunveloper replied:
앗... 이런 대왕실수를 🥹
순서를 보장하려면 `LinkedHashSet ` 을 써야합니다!

geunhwanlee commented on src/main/java/nextstep/subway/domain/Sections.java:
```diff
@@ -0,0 +1,95 @@
+            stationResponses.add(section.getDownStation().toStationResponse());
+        });
+        return stationResponses;
+    }
```
`상행종점역`부터 `하행종점역`까지 순서대로 역을 리턴해주지 못하는것같아요.

임의의 한 section으로부터 상행선 방향으로 탐색을 반복하면 `상행종점역`을 찾을수있고
`상행종점역`으로부터 하행선 방향으로 탐색을 반복하면 정렬된 역을 만들어낼수 있을것 같은데 어떻게 생각하시나요? 😀

geunhwanlee commented on src/main/java/nextstep/subway/domain/Sections.java:
```diff
@@ -0,0 +1,95 @@
+        return changedStationSection == null;
+    }
+
+    public Set<StationResponse> toStationSet() {
```
도메인 모델에서 DTO를 의존하고있는데
`Set<Station>`를 리턴해주고 필요한곳에서 `StationResponse`로 매핑해주는건 어떤가요? 😀

eunveloper replied:
이번 PR에는 전체적으로 도메인모델에서 DTO를 의존하네요 ㅠㅠ
전반적으로 수정 하겠습니다 ㅎ_ㅎ

geunhwanlee reviewed (changes requested):
리뷰사항 반영 잘 해주셨네요! 👍🏻

많은 부분을 변경해야했을텐데 너무 고생 많으셨어요 👏

다만 몇가지 의도와 다르게 동작하는부분이 있어서 수정 요청 드릴게요.. 😭
그 외에도 몇몇 코멘트를 드렸으니 확인해주세요. 🙏

힘드시겠지만 한번 더 수정해주시고 다시 리뷰요청 주세요. 🙇🏻‍♂️
그럼 화이팅입니다! 🔥

eunveloper committed:
refactor: 피드백 반영

- 영속성 전이를 통한 데이터 셋 구조 변경
- 연관관계 주인 설계 반영
- 도메인 모델에 DTO 의존성 제거
- 추가 테스트케이스 개발
- 테스트 코드 리팩토링
- 정렬알고리즘 추가
- 적절한 자료구조로 변경

eunveloper commented:
마지막 댓글에는 리댓글이 달리지 않네요 ㅠㅠ

뭔가 제가 꼼꼼히 못봐서 리뷰어님이 고생하시는것 같아 마음이 불편하네요 🥹 최선을 다해서 본다고 보는데 꼭 누락이 되네요 ㅠㅠ

이번 PR을 통해서 유독 느낀게 많았던거 같아요!
일단, 도메인에 DTO 의존성을 제거해야 한다는것도 느꼈고
제가 JPA 영속성전이를 몰라서 불필요한 로직도 있었고?
아직 매핑구조를 완벽히 숙지를 못하기도 했네요 ㅠㅠ (이번주 스터디때 빡공 해봐야 겠어요!)

테스트코드도 어떤 구조가 좋을지 3주동안 3명의 리뷰어분과 얘기 나누면서
점차 개선이 되는거 같지만 아직 갈길이 멀어보이는 🥲

도메인 모델에 getter 지양은... 머리와 몸이 모두 알고있지만 뜻대로 되지 않게 되네요..ㅠㅠ
꽤나 골머리 썩힌 step3 이었지만 굉장히 많이 배웠어요! 감사합니다 ㅎ_ㅎ

geunhwanlee commented on src/main/java/nextstep/subway/dto/LineResponse.java:
```diff
@@ -1,31 +1,78 @@
         this.color = color;
-        this.stations.add(StationResponse.of(upStation));
-        this.stations.add(StationResponse.of(downStation));
+        this.stations = sortSections(sections);
```
DTO에서 sections를 정렬하는 책임을 가지고있네요. 🤔
sections를 정렬하여 Station의 리스트로 리턴해주는 책임을 다른 클래스에게 위임하고
DTO는 데이터를 받아 클라이언트에게 응답해주는 역할만을 하는게 낫지않을까 생각해요. 😀

geunhwanlee commented on src/main/java/nextstep/subway/dto/LineResponse.java:
```diff
@@ -1,31 +1,78 @@
+        return firstSection;
+    }
+
+    private Section findFirstSection(Sections sections) {
+        Section firstSection = null;
+        for (Section section : sections.getList()) {
+            firstSection = resetFirstSection(section, firstSection);
+        }
+        return firstSection;
+    }
+
+    private Section resetFirstSection(Section section, Section firstSection) {
+        if (section.isPreStation(firstSection)) {
+            return section;
+        }
+        return firstSection;
+    }
```
`상행종점역`을 찾는 로직이 미완성인것 같아요. 😭
`sections.getList()` 의 결과로 `C-D`, `A-B`, `B-C` 순서의 리스트가 있다면
로직의 결과로 `A`가 나와야하지만 `B`가 나올것 같아요!

eunveloper replied:
앗 그러네요 firstSection을 기준으로 앞선 역을 가져올때까지 순회해야하네요! 변경했습니다

geunhwanlee commented on src/main/java/nextstep/subway/dto/LineResponse.java:
```diff
@@ -1,31 +1,78 @@
+
+        Section firstSection = findFirstSection(sections);
+        stationResponses.add(StationResponse.of(firstSection.getUpStation()));
+        for (int i = 0; i < sections.getList().size(); i++) {
+            addedNextSectionStations(firstSection, sections, stationResponses);
+        }
```
`addedNextSectionStations(firstSection, sections, stationResponses);`
구문이 `sections`의 사이즈만큼 반복될거 같네요. 😭
반복문이 빠져야할것 같아요!

eunveloper replied:
이거는 상행종점행 기준으로 다음구간을 가져오고 그 다음구간을 기준으로 또 다음구간을 가져와야해서
보유한 Section 사이즈만큼 반복하며 가져오는게 맞았던거 같아요 ㅠㅠ.. 다른방식이 있을까요?

geunhwanlee replied:
`addedNextSectionStations`에서 `sections`의 사이즈만큼 반복이 돌고있더라구요.
결국 풀어서 쓴다면
```java
for (int i = 0; i < sections.getList().size(); i++) {
   for (Section section : sections.getList()) {
       firstSection = resetNextSection(section, firstSection);
       stationResponses.add(StationResponse.of(firstSection.getUpStation()));
       stationResponses.add(StationResponse.of(firstSection.getDownStation()));
   }
}
```
이렇게 될거같아요!

```java
currentStation = firstStation;
while (currentStation != null) {
    currentStation = nextSection(currentStation, sections);
}
```
힌트를 드리자면 이렇게 하면 되지 않을까요? 😀

다만 이 로직은 시간복잡도가 n*n인데요
HashMap을 사용해서 시간복잡도를 n으로 줄여볼수도 있을것 같네요! 👍🏻
```java
currentStation = firstStation;
while (map.containsKey(currentStation)) {
    currentStation = map.get(currentStation);
}
```

geunhwanlee commented on src/main/java/nextstep/subway/dto/LineResponse.java:
```diff
@@ -1,31 +1,78 @@
+        }
+    }
+
+    private Section resetNextSection(Section section, Section firstSection) {
+        if (section.isPostStation(firstSection)) {
+            return section;
+        }
+        return firstSection;
+    }
```
다음역을 찾아가는 과정도 `firstSection`을 찾는 로직과 마찬가지로 의도대로 동작하지 않을것 같아요.. 😭

eunveloper replied:
수정완료했습니다!

geunhwanlee commented on src/test/java/nextstep/subway/section/SectionAcceptanceTest.java:
```diff
@@ -0,0 +1,242 @@
+        assertThat(response.statusCode()).isEqualTo(HttpStatus.BAD_REQUEST.value());
+    }
+
+    /**
+     * Given 새로운 노선을 생성하고
+     * When 상행역과 하행역과 동일한 구조의 새로운 역을 등록하면
+     * Then 정상적으로 등록되지 않는다.
+     */
+    @DisplayName("상행역과 하행역이 이미 노선에 모두 등록되어 있다면 등록할 수 없다.")
+    @Test
+    void createFailSectionSameStation() {
+        // given
+        Line line = createLine("1호선");
+        addedSections(line, upStation, mediumStation, 5);
+        addedSections(line, mediumStation, downStation, 7);
+
+        // when
+        ExtractableResponse<Response> responseCase1 = createSection(line.getId(), upStation, mediumStation, LESS_THAN_DISTANCE);
+        ExtractableResponse<Response> responseCase2 = createSection(line.getId(), mediumStation, upStation, LESS_THAN_DISTANCE);
+        ExtractableResponse<Response> responseCase3 = createSection(line.getId(), mediumStation, downStation, LESS_THAN_DISTANCE);
+
+        // then
+        assertAll (
+                () -> assertThat(responseCase1.statusCode()).isEqualTo(HttpStatus.BAD_REQUEST.value()),
+                () -> assertThat(responseCase2.statusCode()).isEqualTo(HttpStatus.BAD_REQUEST.value()),
+                () -> assertThat(responseCase3.statusCode()).isEqualTo(HttpStatus.BAD_REQUEST.value())
+        );
+    }
```
```java
// when
ExtractableResponse<Response> responseCase4 = createSection(line.getId(), upStation, downStation, LESS_THAN_DISTANCE);

// then
assertThat(responseCase4.statusCode()).isEqualTo(HttpStatus.BAD_REQUEST.value());
```

위 케이스를 테스트해보면 실패하네요. 😭

각 `section` 별로 `상행역`과 `하행역`을 비교하기때문인데요
전체 `section`에서 `상행역`과 `하행역`이 포함되어있는지 확인해야해요.
`A-B-C` 구간이 있다면 `A-C` 구간은 등록할 수 없어야해요.

geunhwanlee reviewed (approved):
리뷰 반영하시느라 고생 많으셨어요. 😭

이번 단계는 여기서 머지하도록 할게요!

의도대로 동작하지 않는 부분이 몇몇군데 있어서 코멘트 드렸어요.

확인해보시고 다음단계때 고민해보시면 좋을것 같아요! 🙏
